### DISCUSSION TITLE
Issues with Receiving BLE Notifications on ESP32 (Works with Python Bleak Library)

### DISCUSSION THREAD

**Discussion by Ahsan0084** [community/community#77158] [Programming Help]
### Body

Hello GitHub Community 🏷️,

I am working on a project involving BLE communication and facing an issue with the ESP32. I've successfully managed to receive BLE notifications in Python using the bleak library on my PC, but I'm encountering difficulties replicating this functionality on an ESP32 S3 microcontroller. I am seeking guidance or suggestions to resolve this issue.

Context:
I have a BLE peripheral that sends data packets using the Nordic UART Service (NUS). When using the bleak library in Python on my PC, I successfully receive notifications from the BLE peripheral. The notification data looks something like this:

``Notification from 6e400001-b5a3-f393-e0a9-e50e24dcca9e (Handle: 33): Nordic UART Service: bytearray(b'J\x10\x00T\x00\x03P\x19`c\x02\x93\x01\x01\x00\x00\x00\x00\x00\x00K\x10\x00T\x00\x03P\x19`c\x02\x93\x01\x01\x00\x00\x00\x00\x00\x00L\x00\x03\x02\xf7\x03\x9f\xb3\x11\xbc\xbd\x11\xfa\xb8\x02\xfca\x03\x9f\x19M\x01\x11\xb7\xd8\x11\xfa>\xffW\xf3|\x08\x13\xff\x05\xf3\x81\x08PN\x03.\x00\x00\x00\n\xbd\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00O\x10\x00T\x00\x03P\x19`c\x02\x93\x01\x01\x00\x00\x00\x00\x00\x00P\x10\x00T\x00\x03P\x19`c\x02\x93\x01\x01\x00\x00\x00\x00\x00\x00')``

However, when attempting to receive these notifications on an ESP32 using the standard BLE libraries `` #include `<BLEDevice.h>``, ```#include <BLEUtils.h>```, ```#include <BLEScan.h>``` , ``#include <BLEAdvertisedDevice.h>``, the notifications are not being received as expected.

Specifics:

- The BLE peripheral uses the UUID 6e400001-b5a3-f393-e0a9-e50e24dcca9e for the notify characteristic.
- The ESP32 successfully connects to the BLE peripheral and discovers its services and characteristics, and also send the commands successfully.
- The issue arises when trying to receive notifications: the callback function meant to handle these notifications is never triggered.
- The default MTU size on the ESP32 is 23 bytes which i found through `getMTU()`, and I suspect this might be part of the issue since the notification packets are larger (220 bytes).
 ```
if (connected) {
        // Retrieve the negotiated MTU size
        
        uint16_t mtuSize = pClient->getMTU();
        Serial.print("Current MTU size: ");
        Serial.println(mtuSize);
    }
```
- Attempts to increase the MTU size on the ESP32 have not been successful. As i read somewhere to use `BLEDevice::setMTU(220);` right after the  `BLEDevice::init("");`. But it doesnt work
- My code handles the BLE setup, service discovery, and notification registration as per standard procedures. Below is the simple notifycallback function that I am using.

```
static void notifyCallback(BLERemoteCharacteristic* pBLERemoteCharacteristic, uint8_t* pData, size_t length, bool isNotify) {
 Serial.println("Notification Callback Triggered");
  
  Serial.print("Received Data: ");
  for (int i = 0; i < length; i++) {
    Serial.print(pData[i], HEX);
    Serial.print(" ");
  }
  Serial.println();

  }
```
- I have also used  `pNotifyCharacteristic->registerForNotify(notifyCallback);` method, which is successfully executed.
which verify that `pNotifyCharacteristic->canNotify()` returns true before registering for notifications.

Request for Help:

- Has anyone faced a similar issue with BLE notifications on the ESP32?
- Are there known limitations or workarounds when dealing with larger data packets in BLE notifications on the ESP32?

I appreciate any insights or suggestions you can offer on this matter. 

Thank you in advance!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by alihoushngi** [comment]
what is ble
